Notes/Algorithms/Data Structures/Hash-Tables.c

```diff
@@ -1,37 +1,345 @@
 #define SUCCESS 0
 #define ERROR -1
-#define OVERFLOW -2
-#define EMPTY -3
+// Defined limit for a string key length 
+#define LIMIT 10
 
-// Hash struct
-typedef struct Hash {
-    int key;
+// Hash Item struct
+typedef struct Item {
+    char* key;
     int value;
-} Hash;
+} Item;
 
-void print() {
+// Hash Table struct
+typedef struct HashTable {
+    int size;
+    Item** items;
+} HashTable;
 
+
+// returns true if a collision would happen for the initially hashed index for a key/value pair and false if no collision would occurred 
+bool isCollision(HashTable * table, int index) {
+    if (table->items[index] != NULL) {
+        return true;
+    }
+    return false;
 }
 
-void traverse() {
+// equals checks the testHash and expectedHash for parity
+// if the two HashTables match returns true
+// if a mismatch occurs returns false
+bool equals(HashTable * testHash, HashTable * expectedHash) {
+    for (int i = 0; i < expectedHash->size; i++) {
+        if (strcmp(testHash->items[i]->key, expectedHash->items[i]->key) != 0 || testHash->items[i]->value != expectedHash->items[i]->value) {
+            return false;
+        }
+    }
+    return true;
+}
 
+// hashing 
+// returns a "hashed" index based on modulus of the key's ASCII integer sum and the SIZE of the hashTable
+// Explanation:  I used loselose because it was a simple hashing function that seemed adequate for the purposes of this exercise.
+int hashing(HashTable *hashtable, char *key ) {
+    int i = 0;
+    int count = 0;
+    while (key[i] != NULL) {
+        count += key[i];
+        i++;
+    }
+    return count % hashtable->size;
 }
 
-bool compare() {
-    return true;
-    return false;
+
+// creates a new Item struct
+// returns new struct
+Item * createItem(char key[], int value, int * success) {
+    Item * item = malloc(sizeof(Item));
+    if (item != NULL) {        
+        item->key = strdup(key);
+        item->value = value;
+        *success = SUCCESS;
+        return item;
+    }
+    return NULL;
 }
 
+// creates a new HashTable struct
+// returns new HashTable struct
+HashTable * createHashTable(int size) {
+    if (size < 1 || size > 100) return NULL;
+    HashTable * hashTable = malloc(sizeof(HashTable));
+    hashTable->items = malloc(sizeof(Item) * size);
+    if (hashTable == NULL || hashTable->items == NULL) {
+        return NULL;
+    }
+    hashTable->size = size;
+    for (int i = 0; i < size; i++) {
+        hashTable->items[i] = NULL;
+    }
+    return hashTable;
+}
 
 
-// To-dos:
-// Insert a key
-// Delete a key
+// insert 
+// adds the new Item struct to the hashTable
+// if the expected value is in the hashTable at the expected key then returns a success code
+int insert(HashTable * hashTable, char key[], int value) {
+    int success = ERROR;
+    int index = hashing(hashTable, key);
+    while (isCollision(hashTable, index) && index <= hashTable->size) {
+        //find next free space  
+        index++; 
+        index % hashTable->size;  
```

Comment:
It seems like this line is not changing `index`, so not 100% sure what the actual behavior is, can you please elaborate on the desired behavior?

Reply:
Included updates from Walt's comments in my commit to update this function: [commit 13af206b1dc68dab29f360e9ef0f99ffea2c896c](https://github.com/kcarrel/Training-Notes/pull/12/commits/13af206b1dc68dab29f360e9ef0f99ffea2c896c).

My thought process with this function:
- calculates the hashIndex with the given key
- checks to see if the initial hashIndex would result in a collision if the key/value pair was inserted
- If no collision would occur then the key value pair is inserted at the initially calculated hashIndex
-  If a collision does occur then the slots from initially calculated hashIndex to the hashTable->size is checked to see if empty to insert the key/value pair. If no empty slot is found once the end of the hashTable size is reached then a second loop is entered to check from the beginning of the hashTable to the initially calculated hashIndex for an empty slot.